smartapps/smartthings/ecobee-connect.src/ecobee-connect.groovy

```diff
@@ -307,71 +316,101 @@ def getEcobeeDevices() {
 	def thermostatList = [:]
 	def remoteSensors = [:]
 	def switchList = [:]
-	try {
-		// First get thermostats and thermostat remote sensors
-		def bodyParams = [
-				selection: [
-						selectionType:  "registered",
-						selectionMatch: "",
-						includeRuntime: true,
-						includeSensors: true
+	def isThermostatPolled = false
+	def isSwitchesPolled = false
+	def pollAttempt = 1
+	// try obtain devices twice, in case authToken needs to be refreshed
+	while (!(isThermostatPolled && isSwitchesPolled) && (pollAttempt < 3)) {
+		try {
+			// First get thermostats their remote sensors
+			if (!isThermostatPolled) {
+				def bodyParams = [
+						selection: [
+								selectionType:  "registered",
+								selectionMatch: "",
+								includeSettings:        true,
+								includeRuntime:         true,
+								includeSensors:         true
+						]
 				]
-		]
-		def deviceListParams = [
-				uri:     apiEndpoint,
-				path:    "/1/thermostat",
-				headers: ["Content-Type": "text/json", "Authorization": "Bearer ${state.authToken}"],
-				// TODO - the query string below is not consistent with the Ecobee docs:
-				// https://www.ecobee.com/home/developer/api/documentation/v1/operations/get-thermostats.shtml
-				query: [format: 'json', body: toJson(bodyParams)]
-		]
-		httpGet(deviceListParams) { resp ->
-			if (resp.status == 200) {
-				resp.data.thermostatList.each { stat ->
-					def dni = [app.id, stat.identifier].join('.')
-					thermostatList[dni] = getThermostatDisplayName(stat)
-					// compile all remote sensors conected to the thermostat
-					stat.remoteSensors.each { sensor ->
-							if (sensor.type != "thermostat") {
-								def rsDni = "ecobee_sensor-"+ sensor?.id + "-" + sensor?.code
-								remoteSensors[rsDni] = sensor
-								remoteSensors[rsDni] << [thermostatId: dni]
+				def deviceListParams = [
+						uri:     apiEndpoint,
+						path:    "/1/thermostat",
+						headers: ["Content-Type": "text/json", "Authorization": "Bearer ${state.authToken}"],
+						// TODO - the query string below is not consistent with the Ecobee docs:
+						// https://www.ecobee.com/home/developer/api/documentation/v1/operations/get-thermostats.shtml
+						query: [format: 'json', body: toJson(bodyParams)]
+				]
+				httpGet(deviceListParams) { resp ->
+					isThermostatPolled = true
+					if (resp.status == 200) {
+						resp.data.thermostatList.each { stat ->
```

Comment:
should we add the null safety operator? `resp.data?.thermostatList?.each`

Reply:
If we get a 200 ok on the request then the response should have data with a thermostatList, if not there's something seriously wrong that we might want to discover sooner rather than later.

Reply:
👍